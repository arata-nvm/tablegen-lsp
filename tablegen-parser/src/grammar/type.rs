use crate::{
    kind::{SyntaxKind, TokenKind},
    parser::Parser,
    T,
};

use super::value;

pub(super) const TYPE_FIRST_TOKENS: [TokenKind; 8] = [
    T![bit],
    T![int],
    T![string],
    T![dag],
    T![bits],
    T![list],
    T![code],
    TokenKind::Id,
];

// Type ::= BitType | IntType | StringType | DagType | BitsType | ListType | ClassId
pub(super) fn r#type(p: &mut Parser) {
    match p.current() {
        T![bit] => bit_type(p),
        T![int] => int_type(p),
        T![string] => string_type(p),
        T![dag] => dag_type(p),
        T![bits] => bits_type(p),
        T![list] => list_type(p),
        TokenKind::Id => class_id(p),
        _ => p.error_and_recover("unknown token when expecting a type"),
    };
}

// BitType ::= "bit"
pub(super) fn bit_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![bit]);
    p.wrap(m, SyntaxKind::BitType);
}

// IntType ::= "int"
pub(super) fn int_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![int]);
    p.wrap(m, SyntaxKind::IntType);
}

// StringType ::= "string"
pub(super) fn string_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![string]);
    p.wrap(m, SyntaxKind::StringType);
}

// DagType ::= "dag"
pub(super) fn dag_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![dag]);
    p.wrap(m, SyntaxKind::DagType);
}

// BitsType ::= "bits" "<" Integer ">"
pub(super) fn bits_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![bits]);
    p.expect_with_msg(T![<], "expected '<' after bits type");
    value::integer(p).or_error(p, "expected integer in bits<n> type");
    p.expect_with_msg(T![>], "expected '>' at end of bits<n> type");
    p.wrap(m, SyntaxKind::BitsType);
}

// ListType ::= "list" "<" Type ">"
pub(super) fn list_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![list]);
    p.expect_with_msg(T![<], "expected '<' after list type");
    r#type(p);
    p.expect_with_msg(T![>], "expected '>' at end of list<ty> type");
    p.wrap(m, SyntaxKind::ListType);
}

// ClassId ::= Identifier
pub(super) fn class_id(p: &mut Parser) {
    let m = p.marker();
    value::identifier(p).or_error(p, "expected name for ClassID");
    p.wrap(m, SyntaxKind::ClassId);
}
