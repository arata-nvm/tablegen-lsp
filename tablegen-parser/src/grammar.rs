pub mod statement;
pub mod r#type;
pub mod value;

use crate::{
    error::SyntaxError, grammar::statement::StatementListType, language::SyntaxNode,
    parser::Parser, syntax_kind::SyntaxKind, token_kind::TokenKind, T,
};

const RECOVER_TOKENS: [TokenKind; 5] = [T![include], T![class], T![def], T![let], T![;]];

pub fn parse(text: &str) -> (SyntaxNode, Vec<SyntaxError>) {
    let mut parser = Parser::new(text, &RECOVER_TOKENS);
    root(&mut parser);

    let errors = parser.take_errors();
    (parser.finish(), errors)
}

// Root ::= StatementList
fn root(p: &mut Parser) {
    p.start_node(SyntaxKind::Root);
    statement::statement_list(p, StatementListType::TopLevel);
    if !p.eof() {
        p.error("unexpected input at top level");
    }
    p.finish_node();
}

fn delimited<F>(p: &mut Parser, bra: TokenKind, ket: TokenKind, delim: TokenKind, mut parser: F)
where
    F: FnMut(&mut Parser<'_>),
{
    p.expect(bra);
    while !p.at(ket) && !p.eof() {
        parser(p);

        if !p.eat_if(delim) {
            break;
        }
    }
    p.expect(ket);
}

#[cfg(test)]
mod tests {
    use crate::grammar::parse;

    #[test]
    fn statement() {
        insta::assert_debug_snapshot!(parse("a"));
    }

    #[test]
    fn include() {
        insta::assert_debug_snapshot!(parse(r#"include "foo.td""#));

        insta::assert_debug_snapshot!(parse("include"));
    }

    #[test]
    fn class() {
        insta::assert_debug_snapshot!(parse("class Foo<int A, int B = 1>: Bar<A, 2>;"));
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A> {
                int B;
                int C = A;
                let D = A;
            }"
        ));
        insta::assert_debug_snapshot!(parse("class"));
    }

    #[test]
    fn def() {
        insta::assert_debug_snapshot!(parse("def Foo : Bar;"));
        insta::assert_debug_snapshot!(parse("def foo {}"));
    }

    #[test]
    fn r#let() {
        insta::assert_debug_snapshot!(parse("let A = 1, B<1...3> = 0b101 in { class Foo; }"));

        insta::assert_debug_snapshot!(parse("let A = 1"));
        insta::assert_debug_snapshot!(parse("let A = 1 {"));
    }

    #[test]
    fn let_item() {
        insta::assert_debug_snapshot!(parse("let"));
        insta::assert_debug_snapshot!(parse("let A"));
    }

    #[test]
    fn multi_class() {
        insta::assert_debug_snapshot!(parse(
            "multiclass foo {
                def _foo1;
                def _foo2;
            }"
        ));
    }

    #[test]
    fn defm() {
        insta::assert_debug_snapshot!(parse("defm foo : bar;"));
    }

    #[test]
    fn defset() {
        insta::assert_debug_snapshot!(parse(
            "defset list<Base> BaseList = {
                def Foo0 : Base<0>;
                def Foo1 : Base<1>;
            }"
        ));
    }

    #[test]
    fn defvar() {
        insta::assert_debug_snapshot!(parse("defvar i = 0;"));
    }

    #[test]
    fn foreach() {
        insta::assert_debug_snapshot!(parse(
            "foreach i = [0, 1] in {
                def Foo # i : Base<i>;
            }"
        ));
    }

    #[test]
    fn r#if() {
        insta::assert_debug_snapshot!(parse("if true then { class Foo; }"));
    }

    #[test]
    fn assert() {
        insta::assert_debug_snapshot!(parse("assert hoge, \"fuga\";"));
    }

    #[test]
    fn template_arg_decl() {
        insta::assert_debug_snapshot!(parse("class Foo<int"));
    }

    #[test]
    fn class_ref() {
        insta::assert_debug_snapshot!(parse("class Foo : Bar<"));
    }

    #[test]
    fn r#type() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<bit A, int B, string C, dag D, bits<32> E, list<int> F, Bar G>;"
        ));
    }

    #[test]
    fn body() {
        insta::assert_debug_snapshot!(parse("class Foo"));
    }

    #[test]
    fn field_def() {
        insta::assert_debug_snapshot!(parse("class Foo { int"));
        insta::assert_debug_snapshot!(parse("class Foo { int A"));
    }

    #[test]
    fn field_let() {
        insta::assert_debug_snapshot!(parse("class Foo { let"));
        insta::assert_debug_snapshot!(parse("class Foo { let A"));
        insta::assert_debug_snapshot!(parse("class Foo { let A = 1"));
    }

    #[test]
    fn value() {
        insta::assert_debug_snapshot!(parse("class Foo<string A = \"hoge\" # \"fuga\">;"));
    }

    #[test]
    fn inner_value() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A = Hoge.Fuga, bits<2> B = Hoge{0...1}, list<int> C = Hoge[0...1]>;"
        ));
    }

    #[test]
    fn simple_value() {
        insta::assert_debug_snapshot!(parse(
            "class Foo<int A = 1, string B = \"hoge\", bit D = false, int E = ?, bits<2> F = {0, 1}, list<int> G = [1, 2], dag H = (add A:$hoge), int I = A, int J = !add(A, B), int K = !cond(false: 1, true: 2)> {
                code C = [{ true }];
            }"
        ));
    }
}
