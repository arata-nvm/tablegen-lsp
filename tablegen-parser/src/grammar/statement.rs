use crate::{
    kind::{SyntaxKind, TokenKind},
    parser::Parser,
    T,
};

use super::{delimited, r#type, value};

// StatementList ::= Statement*
pub(super) fn statement_list(p: &mut Parser) {
    let m = p.marker();
    while !p.eof() {
        statement(p);
    }
    p.wrap(m, SyntaxKind::StatementList);
}

// Statement ::= Include | Assert | Class | Def | Defm | Defset | Defvar | Foreach | If | Let | MultiClass
pub(super) fn statement(p: &mut Parser) {
    match p.current() {
        T![include] => include(p),
        T![class] => class(p),
        T![def] => def(p),
        T![let] => r#let(p),
        _ => p.error_and_eat("expected class, def, defm, defset, multiclass, let or foreach"),
    }
}

// Include ::= "include" String
pub(super) fn include(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![include]);
    value::string(p).or_error(p, "expected filename after include");
    p.wrap(m, SyntaxKind::Include);
}

// Class ::= "class" Identifier TemplateArgList? RecordBody
pub(super) fn class(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![class]);
    value::identifier(p).or_error(p, "expected class name after 'class' keyword");
    opt_template_arg_list(p);
    record_body(p);
    p.wrap(m, SyntaxKind::Class);
}

// Def ::= "def" Value? RecordBody
pub(super) fn def(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![def]);
    value::value(p);
    record_body(p);
    p.wrap(m, SyntaxKind::Def);
}

// Let ::= "let" LetList "in" ( "{" Statement* "}" | Statement );
pub(super) fn r#let(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![let]);
    let_list(p);
    p.expect_with_msg(T![in], "expected 'in' at end of top-level 'let'");
    if p.eat_if(T!['{']) {
        while !p.at(T!['}']) && !p.eof() {
            statement(p);
        }
        p.expect_with_msg(T!['}'], "expected '}' at end of top-level let command");
    } else {
        statement(p);
    }
    p.wrap(m, SyntaxKind::Let);
}

// LetList ::= LetItem ( "," LetItem )*
pub(super) fn let_list(p: &mut Parser) {
    let m = p.marker();
    while !p.eof() {
        let_item(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.wrap(m, SyntaxKind::LetList);
}

// LetItem ::= Identifier ( "<" RangeList ">" )? "=" Value
pub(super) fn let_item(p: &mut Parser) {
    let m = p.marker();
    value::identifier(p).or_error(p, "expected identifier in let expression");
    p.expect_with_msg(T![=], "expected '=' in let expression");
    value::value(p);
    p.wrap(m, SyntaxKind::LetItem);
}

pub(super) fn opt_template_arg_list(p: &mut Parser) {
    if p.at(T![<]) {
        template_arg_list(p);
    }
}

// TemplateArgList ::= "<" TemplateArgDecl ( "," TemplateArgDecl )* ">"
pub(super) fn template_arg_list(p: &mut Parser) {
    let m = p.marker();
    delimited(p, T![<], T![>], T![,], template_arg_decl);
    p.wrap(m, SyntaxKind::TemplateArgList);
}

// TemplateArgDecl ::= Type Identifier ( "=" Value )?
pub(super) fn template_arg_decl(p: &mut Parser) {
    let m = p.marker();
    r#type::r#type(p);
    value::identifier(p).or_error(p, "expected identifier in declaration");
    if p.eat_if(T![=]) {
        value::value(p);
    }
    p.wrap(m, SyntaxKind::TemplateArgDecl);
}

// RecordBody ::= ParentClassList Body
pub(super) fn record_body(p: &mut Parser) {
    let m = p.marker();
    parent_class_list(p);
    body(p);
    p.wrap(m, SyntaxKind::RecordBody);
}

// ParentClassList ::= ( ":" ClassRef ( "," ClassRef )* )?
pub(super) fn parent_class_list(p: &mut Parser) {
    let m = p.marker();
    if !p.eat_if(T![:]) {
        p.wrap(m, SyntaxKind::ParentClassList);
        return;
    }

    while !p.eof() {
        class_ref(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.wrap(m, SyntaxKind::ParentClassList);
}

// ClassRef ::= Identifier ( "<" ArgValueList? ">" )?
pub(super) fn class_ref(p: &mut Parser) {
    let m = p.marker();
    value::identifier(p);
    if p.eat_if(T![<]) {
        arg_value_list(p);
        p.expect_with_msg(T![>], "expected '>' in template value list");
    }
    p.wrap(m, SyntaxKind::ClassRef);
}

// ArgValueList ::= PositionalArgValueList ","? NamedArgValueList
pub(super) fn arg_value_list(p: &mut Parser) {
    let m = p.marker();
    positional_arg_value_list(p);
    p.wrap(m, SyntaxKind::ArgValueList);
}

// PositionalArgValueList ::= ( Value ( "," Value )* ) ?
pub(super) fn positional_arg_value_list(p: &mut Parser) {
    let m = p.marker();
    while !p.eof() {
        value::value(p);
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.wrap(m, SyntaxKind::PositionalArgValueList);
}

// Body ::= ";" | "{" BodyItem* "}"
pub(super) fn body(p: &mut Parser) {
    let m = p.marker();
    if p.eat_if(T![;]) {
        p.wrap(m, SyntaxKind::Body);
        return;
    }

    p.expect_with_msg(T!['{'], "expected ';' or '{' to start body");
    while !p.at(T!['}']) && !p.eof() {
        if !body_item(p) {
            break;
        }
    }
    p.expect(T!['}']);
    p.wrap(m, SyntaxKind::Body);
}

// BodyItem ::= FieldDef | FieldLet | Defvar | Assert
pub(super) fn body_item(p: &mut Parser) -> bool {
    match p.current() {
        T![let] => field_let(p),
        T![code] | TokenKind::Id => field_def(p),
        _ => return false,
    }
    true
}

// FieldDef ::= ( Type | CodeType ) Identifier ( "=" Value )? ";"
pub(super) fn field_def(p: &mut Parser) {
    let m = p.marker();
    if p.at(T![code]) {
        code_type(p);
    } else {
        r#type::r#type(p);
    }
    value::identifier(p).or_error(p, "expected identifier in declaration");
    if p.eat_if(T![=]) {
        value::value(p);
    }
    p.expect_with_msg(T![;], "expected ';' after declaration");
    p.wrap(m, SyntaxKind::FieldDef);
}

// CodeType ::= "code"
pub(super) fn code_type(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![code]);
    p.wrap(m, SyntaxKind::CodeType);
}

// FieldLet ::= "let" Identitfer ( "{" RangeList "}" )? "=" Value ";"
pub(super) fn field_let(p: &mut Parser) {
    let m = p.marker();
    p.assert(T![let]);
    value::identifier(p).or_error(p, "expected field identifier after let");
    p.expect(T![=]);
    value::value(p).or_error(p, "expected '=' in let expression");
    p.expect_with_msg(T![;], "expected ';' after let expression");
    p.wrap(m, SyntaxKind::FieldLet);
}
