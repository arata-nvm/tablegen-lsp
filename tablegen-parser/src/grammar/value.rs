use crate::{
    kind::{SyntaxKind, TokenKind},
    parser::{Parser, Result, ResultExt},
    T,
};

use super::delimited;

// Value ::= SimpleValue ValueSuffix* | Value "#" Value?
pub(super) fn value(p: &mut Parser) -> Result {
    let m = p.marker();
    simple_value(p)?;

    loop {
        // ValueSuffix ::= RangeSuffix | SliceSuffix | FieldSuffix
        match p.current() {
            T![.] => field_suffix(p)?,
            _ => break,
        }
    }

    p.wrap(m, SyntaxKind::Value);
    Ok(())
}

// FieldSuffix ::= "." Identifier
pub(super) fn field_suffix(p: &mut Parser) -> Result {
    let m = p.marker();
    p.assert(T![.]);
    identifier(p).or_error(p, "expected field identifier after '.'")?;
    p.wrap(m, SyntaxKind::FieldSuffix);
    Ok(())
}

// SimpleValue ::= Integer | String | Code | Boolean | Uninitialized | Bits | List | Dag | Identifier | ClassValue | BangOperator | CondOperator
pub(super) fn simple_value(p: &mut Parser) -> Result {
    match p.current() {
        TokenKind::IntVal => integer(p),
        TokenKind::StrVal => string(p),
        TokenKind::CodeFragment => code(p),
        T![true] | T![false] => boolean(p),
        T![?] => uninitialized(p),
        T!['{'] => bits(p),
        T!['['] => list(p),
        T!['('] => dag(p),
        TokenKind::Id => identifier(p),
        kind if kind.is_bang_operator() => bang_operator(p),
        kind if kind.is_cond_operator() => cond_operator(p),
        _ => {
            p.error_and_eat("unknown token when parsing a value");
            Err(())
        }
    }
}

// Integer ::= INT
pub(super) fn integer(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(TokenKind::IntVal)?;
    p.wrap(m, SyntaxKind::Integer);
    Ok(())
}

// String ::= STRING
pub(super) fn string(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(TokenKind::StrVal)?;
    p.wrap(m, SyntaxKind::String);
    Ok(())
}

// Code ::= CODE
pub(super) fn code(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(TokenKind::CodeFragment)?;
    p.wrap(m, SyntaxKind::Code);
    Ok(())
}

// Boolean ::= "true" | "false"
pub(super) fn boolean(p: &mut Parser) -> Result {
    let m = p.marker();
    if !(p.eat_if(T![true]) || p.eat_if(T![false])) {
        p.error("expected true or false");
        return Err(());
    }
    p.wrap(m, SyntaxKind::Boolean);
    Ok(())
}

// Uninitialized ::= "?"
pub(super) fn uninitialized(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(T![?])?;
    p.wrap(m, SyntaxKind::Uninitialized);
    Ok(())
}

// Bits ::= "{" ValueList "}"
pub(super) fn bits(p: &mut Parser) -> Result {
    let m = p.marker();
    value_list(p, T!['{'], T!['}'])?;
    p.wrap(m, SyntaxKind::Bits);
    Ok(())
}

// List ::= "[" ValueList "]"
pub(super) fn list(p: &mut Parser) -> Result {
    let m = p.marker();
    value_list(p, T!['['], T![']'])?;
    p.wrap(m, SyntaxKind::List);
    Ok(())
}

// ValueList ::= Value ( "," Value )*
pub(super) fn value_list(p: &mut Parser, bra: TokenKind, ket: TokenKind) -> Result {
    let m = p.marker();
    delimited(p, bra, ket, T![,], value)?;
    p.wrap(m, SyntaxKind::ValueList);
    Ok(())
}

// Dag ::= ( DagArg DagArgList? )
pub(super) fn dag(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(T!['('])?;
    if !p.at(TokenKind::Id) {
        p.error("expected identifier in dag init");
        return Err(());
    }
    dagarg(p)?;
    if !p.at(T![')']) {
        dagarg_list(p)?;
    }
    p.expect_with_msg(T![')'], "expected ')' in dag init")?;
    p.wrap(m, SyntaxKind::Dag);
    Ok(())
}

// DagArgList ::= DagArg ( "," DagArg )*
pub(super) fn dagarg_list(p: &mut Parser) -> Result {
    let m = p.marker();
    while !p.eof() {
        dagarg(p)?;
        if !p.eat_if(T![,]) {
            break;
        }
    }
    p.wrap(m, SyntaxKind::DagArgList);
    Ok(())
}

// DagArg ::= Value ( ":" VARNAME ) | VARNAME
pub(super) fn dagarg(p: &mut Parser) -> Result {
    let m = p.marker();
    if p.eat_if(TokenKind::VarName) {
        p.wrap(m, SyntaxKind::DagArg);
        return Ok(());
    }

    value(p)?;
    if p.eat_if(T![:]) {
        var_name(p).or_error(p, "expected variable name in dag literal")?;
    }
    p.wrap(m, SyntaxKind::DagArg);
    Ok(())
}

// VarName ::= VARNAME
pub(super) fn var_name(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(TokenKind::VarName)?;
    p.wrap(m, SyntaxKind::VarName);
    Ok(())
}

// Identifier ::= ID
pub(super) fn identifier(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(TokenKind::Id)?;
    p.wrap(m, SyntaxKind::Identifier);
    Ok(())
}

// BangOperator ::= BANGOP "(" ValueList ")"
pub(super) fn bang_operator(p: &mut Parser) -> Result {
    if !p.current().is_bang_operator() {
        p.error_and_eat("expected bang operator");
        return Err(());
    }

    let m = p.marker();
    p.eat(); // eat bang operator
    delimited(p, T!['('], T![')'], T![,], value)?;
    p.wrap(m, SyntaxKind::BangOperator);
    Ok(())
}

// CondOperator ::= CONDOP "(" CondClause ( "," CondClause )* ")"
pub(super) fn cond_operator(p: &mut Parser) -> Result {
    let m = p.marker();
    p.expect(T![!cond])?;
    delimited(p, T!['('], T![')'], T![,], |p| cond_clause(p))?;
    p.wrap(m, SyntaxKind::CondOperator);
    Ok(())
}

// CondClause ::= Value ":" Value
pub(super) fn cond_clause(p: &mut Parser) -> Result {
    let m = p.marker();
    value(p)?;
    p.expect(T![:])?;
    value(p)?;
    p.wrap(m, SyntaxKind::CondClause);
    Ok(())
}
