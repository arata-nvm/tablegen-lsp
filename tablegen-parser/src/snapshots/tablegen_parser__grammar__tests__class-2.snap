---
source: tablegen-parser/src/grammar.rs
expression: "parse(\"class Foo<int A> {\n                int B;\n                int C = A;\n                let D = A;\n            }\")"
---
(
    Root@0..109
      StatementList@0..109
        Class@0..109
          ClassKw@0..5 "class"
          Whitespace@5..6 " "
          Identifier@6..9
            Id@6..9 "Foo"
          TemplateArgList@9..17
            Less@9..10 "<"
            TemplateArgDecl@10..15
              IntType@10..14
                Int@10..13 "int"
                Whitespace@13..14 " "
              Identifier@14..15
                Id@14..15 "A"
            Greater@15..16 ">"
            Whitespace@16..17 " "
          RecordBody@17..109
            ParentClassList@17..17
            Body@17..109
              LBrace@17..18 "{"
              Whitespace@18..35 "\n                "
              FieldDef@35..58
                IntType@35..39
                  Int@35..38 "int"
                  Whitespace@38..39 " "
                Identifier@39..40
                  Id@39..40 "B"
                Semi@40..41 ";"
                Whitespace@41..58 "\n                "
              FieldDef@58..85
                IntType@58..62
                  Int@58..61 "int"
                  Whitespace@61..62 " "
                Identifier@62..64
                  Id@62..63 "C"
                  Whitespace@63..64 " "
                Equal@64..65 "="
                Whitespace@65..66 " "
                Value@66..67
                  InnerValue@66..67
                    Identifier@66..67
                      Id@66..67 "A"
                Semi@67..68 ";"
                Whitespace@68..85 "\n                "
              FieldLet@85..108
                LetKw@85..88 "let"
                Whitespace@88..89 " "
                Identifier@89..91
                  Id@89..90 "D"
                  Whitespace@90..91 " "
                Equal@91..92 "="
                Whitespace@92..93 " "
                Value@93..94
                  InnerValue@93..94
                    Identifier@93..94
                      Id@93..94 "A"
                Semi@94..95 ";"
                Whitespace@95..108 "\n            "
              RBrace@108..109 "}"
    ,
    [],
)
